PowerShellTools/Intellisense/BraceCompletionController.cs

```diff
@@ -0,0 +1,287 @@
+            //int execResult = NextCommandHandler.Exec(ref pguidCmdGroup, nCmdID, nCmdexecopt, pvaIn, pvaOut);
+            switch (nCmdID)
+            {
+                case (uint)VSConstants.VSStd2KCmdID.TYPECHAR:
+                    var typedChar = Char.MinValue;
+                    typedChar = (char)(ushort)Marshal.GetObjectForNativeVariant(pvaIn);
+                    
+                    // If we processed the typed left brace, no need to pass along the command as the char is already added to the buffer.
+                    if (IsLeftBrace(typedChar))
+                    {
+                        CompleteBrace(typedChar);
+                        SetBraceCompleteState(true);
+                        return VSConstants.S_OK;
+                    }
+                    else if (IsRightBrace(typedChar))
+                    {
+                        if (ProcessTypedRightBrace(typedChar))
+                        {
+                            // If this right brace is typed right after typing left brace,
+                            // we just move the caret to the right side of the right brace and return.
+                            // This means we will not add the typed right brace to the text buffer.
+                            SetBraceCompleteState(false);
+                            return VSConstants.S_OK;
+                        }
+                    }
+                    SetBraceCompleteState(false);
+                    break;
+                case (uint)VSConstants.VSStd2KCmdID.RETURN:
+                    // Return in Repl windows would execute the current command 
+                    if (_textView.TextBuffer.ContentType.TypeName.Equals(ReplConstants.ReplContentTypeName, StringComparison.Ordinal))
+                    {
+                        SetBraceCompleteState(false);
+                        break;
+                    }
+                    if (ProcessReturnKey())
+                    {
+                        SetBraceCompleteState(false);
+                        return VSConstants.S_OK;
+                    }
+                    SetBraceCompleteState(false);
+                    break;
+                case (uint)VSConstants.VSStd2KCmdID.BACKSPACE:
+                    // Return in Repl windows would execute the current command 
+                    if (_textView.TextBuffer.ContentType.TypeName.Equals(ReplConstants.ReplContentTypeName, StringComparison.Ordinal))
+                    {
+                        break;
+                    }
+                    if (ProcessBackspaceKey())
+                    {
+                        SetBraceCompleteState(false);
+                        return VSConstants.S_OK;
+                    }
+                    SetBraceCompleteState(false);
+                    break;
+                case (uint)VSConstants.VSStd2KCmdID.DELETE:                    
+                case (uint)VSConstants.VSStd2KCmdID.UNDO:
+                    SetBraceCompleteState(false);

```

Comment:
Are there any other commands that would cause insertions that we may need to worry about?  Comments/snippits/etc?